File: 2022-05/typescript/61_Rotate_List.ts
function rotateRight(head: ListNode | null, k: number): ListNode | null {
    if (!head) return head
    let len = 1
    let tail = head
    while (tail.next) {
        len++
        tail = tail.next
    }
    let steps = k % len
    if (steps === 0) return head
    steps = len - steps - 1
    let mid = head
    while (steps) {
        steps--
        mid = mid.next
    }
    const newHead = mid.next
    mid.next = null
    tail.next = head
    return newHead
}

File: 2022-05/python/61_Rotate_List.py
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head: return head

        length = 1
        tail = head
        while tail.next:
            length += 1
            tail = tail.next

        steps = k % length
        if steps == 0: return head

        steps = length - steps - 1
        mid = head
        while steps:
            steps -= 1
            mid = mid.next

        new_head = mid.next
        mid.next = None
        tail.next = head
        return new_head